### ISSUE TITLE
Error with auvsi_mp.py and Mission Planner

### ISSUE THREAD

**Issue by s4lvi** [auvsi-suas/interop#209]
Hi, I'm trying to get this year's version of auvsi_mp running in mission planner and I keep running into this, even across different machines: "Error running script expected IList[Byte], got str" I haven't had much luck resolving it, any ideas?

**Comment by prattmic** [comment]
Is there a traceback associated with the call that indicates where it came from?

**Comment by JackHarrington** [comment]
I am getting the same error and there is no traceback associated with the call.

**Comment by pmtischler** [comment]
Can you send screenshots / screencast videos showing the steps to reproduce this?

**Comment by JackHarrington** [comment]
![image](https://cloud.githubusercontent.com/assets/10156110/20247368/5b2aa488-a998-11e6-9f4f-799ea432c486.png)

We simply run auvsi_mp.py in Mission Planner and it returns that error. 

We also found that "INTEROP_URL" is spelled incorrectly (as "INTEOR_URL") on line 7.

**Comment by pmtischler** [comment]
I'll be able to take a look later this week. I don't actually have the
ability to setup and test the program.

When others described how they launched the script it involved copy paste,
but here it appears to be linked.

I also recall others previously had issues with their Git clone settings
not interacting well with Windows.

**Comment by DarkmatterVale** [comment]
How would git clone settings have anything to do with this?

I have debugged the error thoroughly and I believe it comes from issues with the sys.append() calls. I'm not sure why this is happening (using sys.append() in terminal works fine) but whenever I take those statements out of the script it appears to "work" (it crashes with an import error, as it should).

**Comment by pmtischler** [comment]
@DarkmatterVale Can you try adding the interop libs to your global Python PATH and then try the script again without `sys.append()`? If that works, maybe we'll just remove those lines and recommend such in the docs.

In terms of git settings, we've previously had issues with different line terminators and git auto rewriting.

**Comment by DarkmatterVale** [comment]
We installed the interop library via pip and then tested the install (it was successful). However, when we removed the ```sys.append()``` lines we got "Error: Could not load module interop" (or a similar error I do not have the computer in front of me at the moment). It appears Mission Planner does not want to load the local libraries. During debugging the only conclusion I could draw was that the iron python installation was not linking with the same version of standard python installed on the machine, hence why the pip packages were not being found by the script.

Just to make sure we weren't crazy, we tested this on two separate computers and got pretty much identical sets of errors.

Yeah that is the conclusion I came to after I posted the comment. Good thing to check because git settings definitely cause problems occasionally...

**Comment by imritzror** [comment]
Hi, 

I got the following messge while trying to run auvsi_mp.py on mission planner: 


![capture](https://cloud.githubusercontent.com/assets/22326955/20860296/6e0dcf54-b97d-11e6-865e-4fe6113bf120.JPG)

Thanks guys !

**Comment by JackHarrington** [comment]
@imritzror You need to install this module: 
`pip install futures`

**Comment by imritzror** [comment]
Thanks for your reply! 

'pip install futures' is not only for a linux system? 

Any way I've downloaded and installed futures - 3.0.5, but it seems it doesn't solve the problem.

**Comment by pmtischler** [comment]
Good news, I've finally got Windows working on my machine. Soon I'll be able to install Mission Planner and test this out myself.

**Comment by imritzror** [comment]
Great news @pmtischler I would like to know of you get the same error as I got.

Thanks!

**Comment by pmtischler** [comment]
I believe I've figured out the problem. It turns out MissionPlanner's use of IronPython (Microsofts implementation of the Python standard) does something non-standard which breaks most of Python's most-used libraries. `multiprocessing` (used by AsyncClient) and `requests` (used by Client) are example libraries which do not support IronPython (at least not in default form).

During this I attempted to break out AsyncClient and use only Client (at potential loss of throughput). The current error is `sys._getframe` not found, which the web recommends passing `-X:FullFrames` when invoking IronPython to resolve. That might be something we can upstream to MissionPlanner, but it'll take more investigation.

For the async mode, I've found [an issue](https://github.com/IronLanguages/main/issues/1420) which dates back to September and claims to fix the `multiprocessing` issue. Maybe the latest builds are behind this fix. I'm not sure I want to get into custom-compiling IronPython and MIssionPlanner though.

To summarize, IronPython and MissionPlanner are broken to some degree. Afaict, our code *should* work, but our dependencies are broken.

We can investigate cross-project fixes, but maybe we should roll forward to our old approach (local communication to separate process which has interop libs and forwards telemetry to interop server). Sigh.

@prattmic @yogishadixit Thoughts?

**Comment by imritzror** [comment]
Hey , 

Thanks for your reply! I overcome the problem and now I'm getting another error when I'm trying to run the auvsi_mp.py script from my windows machine : 


![image](https://cloud.githubusercontent.com/assets/22326955/21116254/8fdf0d5e-c0bd-11e6-8241-7a25ac06771e.png)

Does anyone know what is the problem? I didn't change the scrip beside change the URL adress.

**Comment by pmtischler** [comment]
@imritzror I haven't seen that error before.

Regarding IronPython, I figured maybe the problem is attempting to use standard Python's packages with IronPython. I installed IronPython and Pip, but then it wouldn't install urllib3 (needed by requests). Found [another issue](https://github.com/IronLanguages/main/issues/1355) with IronPython- doesn't look like that'll work either

**Comment by pmtischler** [comment]
Also just noticed that installing IronPython broke MissionPlanner- had to uninstall IronPython and then repair MissionPlanner.

**Comment by pmtischler** [comment]
Unless we have a better idea, I'll add [xmlrpclib](https://docs.python.org/2.7/library/xmlrpclib.html) back and we'll run 2 processes again.

**Comment by pmtischler** [comment]
Attempted to add xmlrpclib back, and it throws an error on import. This is the same code which previously worked, but now throws an error. From searching the web it seems to be an issue with IronPython doing non-standard stuff with unicode.

Maybe I'm doing something wrong? If not, we've got a problem :/

**Comment by pmtischler** [comment]
Although we didn't want to do this previously, maybe we should consider mavproxy.

**Comment by DarkmatterVale** [comment]
Can you please share the error?

**Comment by pmtischler** [comment]
It's something about expecting bytes, bytes array, and got str or visa
versa.

**Comment by imritzror** [comment]
I'm trying to use and running auvsi_mp.py ,I got the followong error:

![image](https://cloud.githubusercontent.com/assets/22326955/21289120/4c0d42a6-c49d-11e6-8950-e24c767ac8c1.png)

Do you know how to fix it? 

Thanks!

**Comment by DarkmatterVale** [comment]
Has there been any update on this since your last post @pmtischler ?

**Comment by pmtischler** [comment]
Progress was paused due to the holiday period.

We've been debating offline on the best approach. We like mavproxy because
it is independent of frontend and multiplatform (so opportunities for
continuous integration), but we're hesitant to inject a layer between
command and control. Once a decision is made it'll be a bit until we can
implement.

Teams are free to make their own integration in the meantime.

**Comment by pmtischler** [comment]
Sufficient time has elapsed and a viable alternative has not presented itself. I'll begin implementing a mavproxy integration.

**Comment by pmtischler** [comment]
Implementation based on tee-d traffic from MavProxy sent for review via #216.

**Comment by DarkmatterVale** [comment]
@pmtischler I'm not sure why this works, but I was able to import xmlrpclib back into Mission Planner's script using the following link:

http://stackoverflow.com/questions/1371994/importing-external-module-in-ironpython

If possible, could you verify that importing works again on the newest version of mission planner?

I apologize for the inconvenience and understand if you are not able to accommodate my request.

**Comment by pmtischler** [comment]
@DarkmatterVale We've moved away from supporting Mission Planner's scripting interface. We now support Mavproxy instead. Integrating with Mavproxy (well, Mavlink streams directly) means we support more ground stations and have smoke tests running in continuous integration (Travis). We're unlikely to move back, short of discovering problems with the Mavproxy approach.

**Comment by DarkmatterVale** [comment]
Ah, ok.
